### PR TITLE
Support review pr using chatgpt

### PR DESCRIPTION
**Pull request by neuyilan**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,3 +16,4 @@
 coverage.txt
 codegpt
 dist
+.idea/
```
Comment (neuyilan) [inline .gitignore:19]: Error generating comment for file .gitignore: Post "https://api.openai.com/v1/chat/completions": proxyconnect tcp: dial tcp 127.0.0.1:10838: connectex: No connection could be made because the target machine actively refused it.

File: cmd/cmd.go [hunk 2]
```diff
@@ -99,6 +99,7 @@ func initConfig() {
 			fmt.Fprintln(os.Stderr, err)
 		}
 	}
+	fmt.Println("Using config file:", viper.ConfigFileUsed())
 }
```
Comment (neuyilan) [inline cmd/cmd.go:104]: Error generating comment for file cmd/cmd.go: Post "https://api.openai.com/v1/chat/completions": proxyconnect tcp: dial tcp 127.0.0.1:10838: connectex: No connection could be made because the target machine actively refused it.

File: cmd/config.go [hunk 3]
```diff
@@ -58,6 +58,12 @@ func init() {
...
 	_ = viper.BindPFlag("git.template_string", configCmd.PersistentFlags().Lookup("template_string"))
+	_ = viper.BindPFlag("git.api_base_url", configCmd.PersistentFlags().Lookup("api_base_url"))
+	_ = viper.BindPFlag("git.token", configCmd.PersistentFlags().Lookup("token"))
+	_ = viper.BindPFlag("git.owner", configCmd.PersistentFlags().Lookup("owner"))
+	_ = viper.BindPFlag("git.repo", configCmd.PersistentFlags().Lookup("repo"))
+	_ = viper.BindPFlag("server.ip", configCmd.PersistentFlags().Lookup("ip"))
+	_ = viper.BindPFlag("server.port", configCmd.PersistentFlags().Lookup("port"))
 }
 
 var configCmd = &cobra.Command{
```
Comment (neuyilan) [inline cmd/config.go:69]: Error generating comment for file cmd/config.go: Post "https://api.openai.com/v1/chat/completions": proxyconnect tcp: dial tcp 127.0.0.1:10838: connectex: No connection could be made because the target machine actively refused it.

File: main.go [hunk 4]
```diff
@@ -1,33 +1,35 @@
...
+	log.SetFlags(log.Lshortfile | log.LstdFlags)
+	cmd.InitConfig()
+	http.HandleFunc("/payload", webhook.HandleWebhook)
+	ip := viper.GetString("server.ip")
+	port := viper.GetString("server.port")
+	var address = fmt.Sprintf("%s:%s", ip, port)
+	fmt.Println("Listening for GitHub Webhooks on", address)
+	http.ListenAndServe(address, nil)
+
 }
```
Comment (neuyilan) [inline main.go:35]: Error generating comment for file main.go: Post "https://api.openai.com/v1/chat/completions": proxyconnect tcp: dial tcp 127.0.0.1:10838: connectex: No connection could be made because the target machine actively refused it.

File: util/util.go [hunk 5]
```diff
@@ -1,9 +1,37 @@
...
+
+		}
+	}
+
+	return "", errors.New("can not find the client ip addresses")
+}
+
+func CountWords(s string) int {
+	return len(strings.Fields(s))
+}
```
Comment (neuyilan) [inline util/util.go:37]: Error generating comment for file util/util.go: Post "https://api.openai.com/v1/chat/completions": proxyconnect tcp: dial tcp 127.0.0.1:10838: connectex: No connection could be made because the target machine actively refused it.

File: webhook/pr_review.go [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
```
Comment (neuyilan) [inline webhook/pr_review.go:198]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": proxyconnect tcp: dial tcp 127.0.0.1:10838: connectex: No connection could be made because the target machine actively refused it.

File: webhook/pr_review.go [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Error submitting review comment:", err)
+		return err
+	}
+	defer resp.Body.Close()
+
```
Comment (neuyilan) [inline webhook/pr_review.go:366]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": proxyconnect tcp: dial tcp 127.0.0.1:10838: connectex: No connection could be made because the target machine actively refused it.

File: webhook/pr_review.go [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+	if resp.StatusCode == http.StatusCreated {
+		log.Println("Review comment submitted successfully")
+	} else {
+		return fmt.Errorf("error: %d", resp.StatusCode)
+	}
+
+	return nil
+}
```
Comment (neuyilan) [inline webhook/pr_review.go:374]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": proxyconnect tcp: dial tcp 127.0.0.1:10838: connectex: No connection could be made because the target machine actively refused it.

File: .gitignore (hunk 1 above)
Comment (neuyilan) [inline .gitignore:19]: It seems that the provided code patch is incomplete and only contains a single line of code. Therefore, it is difficult to conduct a comprehensive code review. 

However, based on the provided code patch, it only adds the file/folder names to the coverage.txt file, excludes the codegpt folder, and includes the dist and .idea folders. This change does not appear to introduce any bug risk or security vulnerabilities.

As for improvement suggestions, it would be beneficial to provide more context and information about the codebase and the purpose of this change. Additionally, it is recommended to follow a consistent coding style and use descriptive variable and function names to enhance code readability and maintainability.

File: cmd/cmd.go (hunk 2 above)
Comment (neuyilan) [inline cmd/cmd.go:104]: Error generating comment for file cmd/cmd.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: cmd/config.go (hunk 3 above)
Comment (neuyilan) [inline cmd/config.go:69]: Error generating comment for file cmd/config.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: main.go (hunk 4 above)
Comment (neuyilan) [inline main.go:35]: Error generating comment for file main.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: util/util.go (hunk 5 above)
Comment (neuyilan) [inline util/util.go:37]: Error generating comment for file util/util.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: webhook/pr_review.go (hunk 6 above)
Comment (neuyilan) [inline webhook/pr_review.go:198]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: webhook/pr_review.go (hunk 7 above)
Comment (neuyilan) [inline webhook/pr_review.go:366]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: webhook/pr_review.go (hunk 8 above)
Comment (neuyilan) [inline webhook/pr_review.go:374]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: .gitignore (hunk 1 above)
Comment (neuyilan) [inline .gitignore:19]: Error generating comment for file .gitignore: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: cmd/cmd.go [hunk 9]
```diff
@@ -99,6 +100,7 @@ func initConfig() {
 			fmt.Fprintln(os.Stderr, err)
 		}
 	}
+	fmt.Println("Using config file:", viper.ConfigFileUsed())
```
Comment (neuyilan) [inline cmd/cmd.go:103]: Error generating comment for file cmd/cmd.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: cmd/config.go (hunk 3 above)
Comment (neuyilan) [inline cmd/config.go:69]: Error generating comment for file cmd/config.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: cmd/review.go [hunk 10]
```diff
@@ -73,6 +74,7 @@ var reviewCmd = &cobra.Command{
 
 		// Get summarize comment from diff datas
 		color.Cyan("We are trying to review code changes")
+		fmt.Println("11111111111111, ", cmd.Context())
 		resp, err := client.Completion(cmd.Context(), out)
 		if err != nil {
```
Comment (neuyilan) [inline cmd/review.go:79]: Error generating comment for file cmd/review.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: main.go (hunk 4 above)
Comment (neuyilan) [inline main.go:35]: Overall, the code patch looks clean and well-structured. However, there are a few suggestions for improvement:

1. Use a logger instead of fmt.Println for logging messages. This allows for more flexibility in configuring log levels and output destinations.

2. Add error handling for the http.ListenAndServe function. If there is an error starting the server, the program should exit gracefully instead of crashing.

3. Consider using a library for handling signals instead of manually setting up a signal channel. Libraries like os/signal provide more robust and cross-platform signal handling.

4. Add comments to the code to explain the purpose and logic of each function and block of code. This makes it easier for future developers to understand and maintain the code.

5. Use constants or configuration variables instead of hard-coding values like "/payload" and "server.ip" and "server.port". This makes the code more modular and easier to configure for different environments.

File: util/util.go (hunk 5 above)
Comment (neuyilan) [inline util/util.go:37]: Code review:

1. The import statement is repeated twice, which can be consolidated into a single import statement.
2. The IsCommandAvailable() function checks whether a given command is available in the system or not, which is a good practice to prevent errors during runtime.
3. The GetClientIp() function retrieves the IP address of the client, but it only returns the first IPv4 address it encounters. This may not be the intended behavior and could lead to unexpected results.
4. The CountWords() function counts the number of words in a given string, which is a useful utility function.

Improvement suggestions:
1. Consolidate the import statement into a single line.
2. Consider adding error handling when retrieving the IP address to handle edge cases where multiple IP addresses are present or when the IP address is not retrievable for some reason.
3. Consider adding unit tests for each function to ensure that they are working as intended.

File: webhook/pr_review.go (hunk 6 above)
Comment (neuyilan) [inline webhook/pr_review.go:198]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: webhook/pr_review.go (hunk 7 above)
Comment (neuyilan) [inline webhook/pr_review.go:366]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: webhook/pr_review.go (hunk 8 above)
Comment (neuyilan) [inline webhook/pr_review.go:374]: As an AI language model, I cannot determine the context of the code patch. However, based on the syntax, it seems to be a simple HTTP request to create a review comment. 

In terms of security vulnerabilities, it is always recommended to use HTTPS instead of HTTP to encrypt the data being transferred. Also, it is important to ensure that the request is being made to a trusted endpoint and that any user input is properly sanitized to prevent injection attacks. 

Regarding improvement suggestions, it is good practice to include error handling for any potential errors that may occur during the request. Additionally, it may be useful to include more descriptive error messages to help with debugging in case of failures.

File: .gitignore (hunk 1 above)
Comment (neuyilan) [inline .gitignore:19]: This code patch appears to be a simple modification to a list of files to be excluded from code coverage analysis and distribution packaging. As such, there do not appear to be any bug risks or security vulnerabilities associated with this code.

One suggestion for improvement could be to add comments to the code to explain the purpose of each file or directory being excluded. This can help future developers understand the reasoning behind the exclusions and ensure that they are not accidentally removed or modified.

File: cmd/cmd.go (hunk 9 above)
Comment (neuyilan) [inline cmd/cmd.go:103]: Error generating comment for file cmd/cmd.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: cmd/config.go (hunk 3 above)
Comment (neuyilan) [inline cmd/config.go:69]: It's difficult to provide a comprehensive code review with only a small snippet of code, but there are a few things that can be commented on:

- It's good to see that the code is using viper to bind command-line flags to configuration parameters. This can make it easier to manage configuration options.
- It's unclear what the purpose of the `configCmd` variable is without seeing the rest of the code. However, assuming that it is a command in a CLI application, it's good to see that it is using Cobra to manage the command-line interface.
- It's unclear what the purpose of the `init()` function is without seeing the rest of the code. However, it's worth noting that `init()` functions can be difficult to reason about and can lead to unexpected behavior. It's generally better to explicitly call initialization functions rather than relying on `init()` functions.
- There is no obvious security vulnerability in the code snippet provided, but it's impossible to say for certain without seeing the rest of the code.
- It's difficult to provide improvement suggestions without more context, but it's generally a good idea to make sure that configuration options are validated and sanitized to prevent unexpected behavior. It's also a good idea to make sure that any sensitive configuration options (such as API tokens) are properly secured.

File: cmd/review.go (hunk 10 above)
Comment (neuyilan) [inline cmd/review.go:79]: Overall, the code patch looks relatively simple and straightforward. However, there are a few points to consider:

- The code patch seems to be missing some context. It's unclear what package and libraries are being used, what the `Completion` function does, and what `out` represents.
- There is a print statement (`fmt.Println`) with a hardcoded string (`"11111111111111, "`) that seems unnecessary and could be removed.
- There is no error handling for the `color.Cyan` statement, which could potentially cause issues if the function fails. It's always good to have error handling for any critical statements.
- It's unclear what the purpose of this command is and how it fits into the overall application. Adding some comments or documentation to explain this would be helpful for anyone reviewing the code.

File: main.go (hunk 4 above)
Comment (neuyilan) [inline main.go:35]: Error generating comment for file main.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: util/util.go (hunk 5 above)
Comment (neuyilan) [inline util/util.go:37]: Error generating comment for file util/util.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: webhook/pr_review.go (hunk 6 above)
Comment (neuyilan) [inline webhook/pr_review.go:198]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

File: webhook/pr_review.go (hunk 7 above)
Comment (neuyilan) [inline webhook/pr_review.go:366]: Overall, the code patch looks well-written and structured. However, here are some suggestions for improvement:

1. Error handling: The code handles errors using log statements, which are useful for debugging but not for notifying the user of errors. Consider returning errors to the caller or using panic for fatal errors.

2. Security: The code uses an API token for authentication, which could be a security risk if the token is leaked. Consider using environment variables or a secure storage system for storing secrets.

3. Code quality: The code could benefit from more comments and documentation to help future maintainers understand the purpose and functionality of each function.

4. Efficiency: The code makes multiple HTTP requests to the same endpoint. Consider caching responses or using a library that supports connection pooling to improve performance.

File: webhook/pr_review.go (hunk 8 above)
Comment (neuyilan) [inline webhook/pr_review.go:374]: Error generating comment for file webhook/pr_review.go: Post "https://api.openai.com/v1/chat/completions": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
